providers/apis/dydx/switch_over_fetcher.go

```diff
@@ -0,0 +1,153 @@
+func DefaultSwitchOverMarketMapFetcher(
+	logger *zap.Logger,
+	api config.APIConfig,
+	rh apihandlers.RequestHandler,
+	metrics metrics.APIMetrics,
+) (mmclient.MarketMapFetcher, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger is nil")
+	}
+	if api.Name != SwitchOverAPIHandlerName {
+		return nil, fmt.Errorf("expected api name %s, got %s", SwitchOverAPIHandlerName, api.Name)
+	}
+	if len(api.Endpoints) != 2 {
+		return nil, fmt.Errorf(
+			"expected two endpoints, got %d; the first end point must be the rest port second endpoint gRPC",
+			len(api.Endpoints),
+		)
+	}
+	if rh == nil {
+		return nil, fmt.Errorf("request handler is nil")
+	}
+	if metrics == nil {
+		return nil, fmt.Errorf("metrics is nil")
+	}
+
+	// Construct the dYdX x/prices API handler.
+	pricesAPIHandler, err := NewAPIHandler(logger, api)
+	if err != nil {
+		return nil, err
+	}
+	pricesFetcher, err := apihandlers.NewRestAPIFetcher(
+		rh,
+		pricesAPIHandler,
+		metrics,
+		api,
+		logger,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the dYdX x/marketmap API handler.
+	conn, err := slinkygrpc.NewClient(
+		api.Endpoints[1].URL,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithNoProxy(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	marketmapClient, err := marketmap.NewGRPCClientWithConn(conn, api, metrics)
+	if err != nil {
+		return nil, err
+	}
+	marketmapFetcher, err := marketmap.NewMarketMapFetcherWithClient(logger, marketmapClient)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SwitchOverFetcher{
+		logger:           logger,
+		pricesFetcher:    pricesFetcher,
+		marketmapFetcher: marketmapFetcher,
+	}, nil
+}
+
+// NewSwitchOverFetcher returns a new SwitchOverFetcher with the provided
+// fetchers.
+func NewSwitchOverFetcher(
+	logger *zap.Logger,
+	pricesFetcher mmclient.MarketMapFetcher,
+	marketmapFetcher mmclient.MarketMapFetcher,
+) (mmclient.MarketMapFetcher, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger is nil")
+	}
+
+	if pricesFetcher == nil {
+		return nil, fmt.Errorf("prices fetcher is nil")
+	}
+
+	if marketmapFetcher == nil {
+		return nil, fmt.Errorf("marketmap fetcher is nil")
+	}
+
+	return &SwitchOverFetcher{
+		logger:           logger,
+		pricesFetcher:    pricesFetcher,
+		marketmapFetcher: marketmapFetcher,
+	}, nil
+}
+
+// Fetch fetches the market map from the x/prices API and then the x/marketmap
+// API. The fetcher will switch over to the x/marketmap API the first time it
+// receives a non-nil market map from the x/marketmap API.
+func (f *SwitchOverFetcher) Fetch(
+	ctx context.Context,
+	chains []mmclient.Chain,
+) mmclient.MarketMapResponse {
+	if f.switched {
+		return f.marketmapFetcher.Fetch(ctx, chains)
+	}
+
+	f.logger.Debug("fetching marketmap from x/marketmap")
+	resp := f.marketmapFetcher.Fetch(ctx, chains)
+	if len(resp.Resolved) > 0 {
+		f.logger.Info("got response from x/marketmap; switching over to x/marketmap")
+		f.switched = true
```

Comment:
We'll probs want to know when nodes have switched over to the x/prices market-map. Will there be an easy way to do this?

Reply:
yes, the marketmapclient has metrics associated with it, specifically RPC metrics. So we should be able to see when the switch over happens.